aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,948 @@
+            let next_node = bucket_table::borrow(&tree.nodes, next_index);
+            iter.node_index = next_index;
+            iter.key_index = 0;
+            iter.key = vector::borrow(&next_node.children, 0).max_key;
+            return iter
+        };
+
+        new_end_iter(tree)
+    }
+
+    /// Returns the previous iterator, or none if already at the begin iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun prev_iter<V>(tree: &SmartTree<V>, iter: Iterator): Option<Iterator> {
+        if (iter.node_index == tree.min_leaf_index && iter.key_index == 0) {
+            return option::none()
+        };
+
+        option::some(prev_iter_or_die(tree, iter))
+    }
+
+    /// Returns the previous iterator, aborts if already at the begin iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun prev_iter_or_die<V>(tree: &SmartTree<V>, iter: Iterator): Iterator {
+        let node_index = iter.node_index;
+
+        let prev_index;
+
+        if (node_index == NULL_INDEX) {
+            prev_index = tree.max_leaf_index;
+        } else {
+            let node = bucket_table::borrow(&tree.nodes, node_index);
+            if (iter.key_index >= 1) {
+                iter.key_index = iter.key_index - 1;
+                iter.key = vector::borrow(&node.children, iter.key_index).max_key;
+                return iter
+            };
+            prev_index = node.prev;
+        };
+
+        assert!(prev_index != NULL_INDEX, E_UNKNOWN);
+
+        let prev_node = bucket_table::borrow(&tree.nodes, prev_index);
+        let len = vector::length(&prev_node.children);
+        iter.node_index = prev_index;
+        iter.key_index = len - 1;
+        iter.key = vector::borrow(&prev_node.children, len - 1).max_key;
+        iter
+    }
+
+    //////////////////////////////
+    // Internal Implementations //
+    //////////////////////////////
+
+    fun new_node(is_leaf: bool, parent: u64): Node {
+        Node {
+            is_leaf: is_leaf,
+            parent: parent,
+            children: vector::empty(),
+            prev: NULL_INDEX,
+            next: NULL_INDEX,
+        }
+    }
+
+    fun new_child(max_key: u64, index: u64): Child {
+        Child {
+            max_key: max_key,
+            index: index,
+        }
+    }
+
+    fun new_iter(node_index: u64, key_index: u64, key: u64): Iterator {
+        Iterator {
+            node_index: node_index,
+            key_index: key_index,
+            key: key,
+        }
+    }
+
+    fun find_leaf<V>(tree: &SmartTree<V>, key: u64): u64 {
+        let current = tree.root;
+        while (current != NULL_INDEX) {
+            let node = bucket_table::borrow(&tree.nodes, current);
+            if (node.is_leaf) {
+                return current
+            };
+            let len = vector::length(&node.children);
+            assert!(len != 0, E_UNKNOWN);
```

Comment:
Related to my comment earlier in using the presence of children to decide if a node is a leaf node. This would eliminate the need for this assertion.

Reply:
ditto for why I store is_leaf.

Removed this assert since it doesn't provide extra useful information.